pkg/webhook/resources/persistentvolumeclaim/mutator.go

```diff
@@ -47,32 +48,86 @@ func (m *pvcMutator) Create(_ *types.Request, newObj runtime.Object) (types.Patc
 	pvc := newObj.(*corev1.PersistentVolumeClaim)
 
 	logrus.Debugf("create PVC %s/%s with mutator", pvc.Namespace, pvc.Name)
+	patchOPs := []string{}
+	isGoldenImage := false
 
 	// check pvc is related to the vm image
-	_, err := m.vmImageCache.Get(pvc.Namespace, pvc.Name)
+	patchOp, err := m.patchGoldenImageAnnotation(pvc)
 	if err != nil {
+		return nil, err
+	}
+	if patchOp != "" {
+		isGoldenImage = true
+		patchOPs = append(patchOPs, patchOp)
+	}
+
+	// patch populator
+	if !isGoldenImage {
+		patchOp, err = m.patchDataSource(pvc)
+		if err != nil {
+			return nil, err
+		}
+		patchOPs = append(patchOPs, patchOp)
+	}
+
+	// no need to patch
+	if len(patchOPs) == 0 {
+		return nil, nil
+	}
+	return patchOPs, nil
+}
+
+func (m *pvcMutator) patchGoldenImageAnnotation(pvc *corev1.PersistentVolumeClaim) (string, error) {
+	if _, err := m.vmImageCache.Get(pvc.Namespace, pvc.Name); err != nil {
 		if apierrors.IsNotFound(err) {
 			logrus.Infof("PVC %s/%s is not related to the VM image, skip patch", pvc.Namespace, pvc.Name)
-			return nil, nil
+			return "", nil
 		}
-		return nil, err
+		return "", err
 	}
 
 	annotations := pvc.GetAnnotations()
 
-	if _, find := annotations[util.AnnotationGoldenImage]; find {
-		if annotations[util.AnnotationGoldenImage] == "true" {
-			return nil, nil
-		}
+	if v, find := annotations[util.AnnotationGoldenImage]; find && v == "true" {
+		return "", nil
```

Comment:
This will trigger `m.patchDataSource(pvc)` later. Do we need data source for the golden image?

Reply:
No, this is used for gold images. That path uses the datavolume instead of the populator.

Reply:
To confirm, returning an empty string here will set `isGoldenImage` to `false`, even though the PVC is a golden image. Then the PVC will be patched with the `blank` data source. I guess I am not sure whether the golden image is supposed to have the `blank` data source or not.

Reply:
In theory, it should not have happened because we only do this on the create. And the goldenImage is also controlled by us.
But you were correct: it will happen if users add the goldenImage flag themselves.
I will enhance this part.